smart_card_connector_app/src/pcsc-api-jstocxxtest.js

```diff
@@ -471,6 +471,122 @@ goog.exportSymbol('testPcscApi', {
+          context, API.INFINITE,
+          [new API.SCARD_READERSTATE_IN(
+              SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0,
+              API.SCARD_STATE_EMPTY)]);
+      // Simulate card insertion.
+      setSimulatedDevices([
+        {'id': 123, 'type': 'gemaltoPcTwinReader', 'cardType': 'cosmoId70'}
+      ]);
+      const result = await resultPromise;
+
+      let readerStates = null;
+      result.get(
+          (readerStatesArg) => {
+            readerStates = readerStatesArg;
+          },
+          (errorCode) => {
+            fail(`Unexpected error ${errorCode}`);
+          });
+      assertEquals(readerStates.length, 1);
+      const state = readerStates[0];
+      assertEquals(
+          state['reader_name'],
+          SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0);
+      assertEquals(state['current_state'], API.SCARD_STATE_EMPTY);
+      // The "event_state" field contains the number of card insertion/removal
+      // events in the higher 16 bits.
+      assertEquals(
+          state['event_state'],
+          API.SCARD_STATE_CHANGED | API.SCARD_STATE_PRESENT | 0x10000);
+      assertObjectEquals(state['atr'], SimulationConstants.COSMO_ID_70_ATR);
+      assertEquals(result.getErrorCode(), API.SCARD_S_SUCCESS);
+    },
+
+    // Test `SCardGetStatusChange()` detects when a card is removed.
+    'testSCardGetStatusChange_cardRemoving': async function() {
+      // Start with a connected reader and a card.
+      await launchPcscServer(
+          /*initialDevices=*/[
+            {'id': 123, 'type': 'gemaltoPcTwinReader', 'cardType': 'cosmoId70'}
+          ]);
+      const context = await establishContextOrThrow();
+
+      const resultPromise = client.api.SCardGetStatusChange(
+          context, API.INFINITE,
+          [new API.SCARD_READERSTATE_IN(
+              SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0,
+              API.SCARD_STATE_PRESENT)]);
+      // Simulate the card removal.
+      setSimulatedDevices([{'id': 123, 'type': 'gemaltoPcTwinReader'}]);
+      const result = await resultPromise;
+
+      let readerStates = null;
+      result.get(
+          (readerStatesArg) => {
+            readerStates = readerStatesArg;
+          },
+          (errorCode) => {
+            fail(`Unexpected error ${errorCode}`);
+          });
+      assertEquals(readerStates.length, 1);
+      const state = readerStates[0];
+      assertEquals(
+          state['reader_name'],
+          SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0);
+      assertEquals(state['current_state'], API.SCARD_STATE_PRESENT);
+      // The "event_state" field contains the number of card insertion/removal
+      // events in the higher 16 bits.
+      assertEquals(
+          state['event_state'],
+          API.SCARD_STATE_CHANGED | API.SCARD_STATE_EMPTY | 0x10000);
+      assertObjectEquals(state['atr'], new ArrayBuffer(0));
+      assertEquals(result.getErrorCode(), API.SCARD_S_SUCCESS);
```

Comment:
Can you put this in a separate function maybe?
```
 let readerStates = null;
      result.get(
          (readerStatesArg) => {
            readerStates = readerStatesArg;
          },
          (errorCode) => {
            fail(`Unexpected error ${errorCode}`);
          });
      assertEquals(readerStates.length, 1);
      const state = readerStates[0];
      assertEquals(
          state['reader_name'],
          SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0);
      assertEquals(state['current_state'], API.SCARD_STATE_PRESENT);
      // The "event_state" field contains the number of card insertion/removal
      // events in the higher 16 bits.
      assertEquals(
          state['event_state'],
          API.SCARD_STATE_CHANGED | API.SCARD_STATE_EMPTY | 0x10000);
      assertObjectEquals(state['atr'], new ArrayBuffer(0));
      assertEquals(result.getErrorCode(), API.SCARD_S_SUCCESS);
  ```
  Like `CheckGetStatusChangeResult(reader_name, current_state, event_state, atr);`

Reply:
Done, good idea. Originally I checked whether Closure Library allows GMock-style matchers, but haven't found it. But your proposal with a helper function is good enough.